Rewrite of the Getting Started guide (#162)
projectsyn/documentation: docs/modules/ROOT/pages/tutorials/getting-started-2.adoc

> = Getting Started with Project Syn (Part 2)

Reviewer: Part 2 is not linked in the navigation as far as I can tell.

Author: It will not. It is not meant as a standalone document, but rather a followup of the first part.

Reviewer 2: But then why split it up if it is not standalone? "Because the document gets too long" IMHO is a lame argument.

Author: It is **precisely** the argument. Length is a natural deterrent in any learning process. Besides, there will be a "Project Syn the Hard Way™®©" followup tutorial with much more in-depth. 

I refer to the original ticket that gave birth to this PR: 
As a CTO
I want to have a first impression in less than 30 minutes
So that I can evaluate Syn

Author: I also refer to research I'm reading about learning material design: https://twitter.com/akosma/status/1543530358670606340

> = Getting Started with Project Syn (Part 2)

Reviewer 2: Can you add subtitles instead of arbitrary numbering? The subtitle will give me a better understanding on what awaits me and how the two (maybe more in future) getting started differ from each other.

You can use an admonition to point out that this is supposed to be read after the other one.

Author: What do you mean with subtitles?

Reviewer 2: Conceptually, a subtitle can be something like "Getting Started - Set up a lab", "Getting Started - working with cluster catalogs", or something along that line.

Technically: https://docs.asciidoctor.org/asciidoc/latest/document/subtitle/. But this is not what I had in mind. On HTML output, it is ignored anyway. The conceptual thing above is more important.

Author: Done

== Requirements

If you haven't yet, please execute the steps in the xref:tutorials/gettings-started.adoc["Getting Started with Project Syn"] tutorial before attempting the instructions in this document.

This tutorial assumes that you have the "k3d-lieutenant" and "k3d-steward" clusters up and running and that you have installed all required tools.

In addition, create the following environment variable with the path of your private key:

[source,bash]
--
export COMMODORE_SSH_PRIVATE_KEY=~/.ssh/id_ed25519
--

Commodore requires this variable to be able to push changes to the GitLab repositories on your behalf.

== Commodore

Commodore is the third component of the Project Syn suite of tools. It's a command-line tool that compiles the state of your clusters and pushes the changes to the GitLab projects. Steward watches those projects continuously and immediately applies the changes to your clusters, following the GitOps philosophy.

Compile the information of your `k3d-steward` cluster using Commodore by running the command below, which will create a few environment variables in your shell.

[source,bash]
----
source <(curl -fsL https://try.syn.tools/commodore)

> ----

Reviewer: Might be nice to have a info box here explaining what the command does and point to https://syn.tools/commodore/explanation/running-commodore.html (which should be moved to how-to probably) for people who are interested in setting up a local Commodore more permanently

Author: Done

After the command above, you should be able to see a few variables and functions in your shell, for example, the Project Syn ID of your "k3d-steward" cluster:

[source,bash]
----
echo $CLUSTER_ID
----

The output of this command should be a string similar to "c-floral-wood-9134."

Let's now execute Commodore and compile the cluster.

[source,bash]
----
commodore catalog compile $CLUSTER_ID
----

Let's this time compile and push the cluster to GitLab.

[source,bash]
----
commodore catalog compile $CLUSTER_ID --push
----

> At this point, there should be a GitLab project under your profile called `project-syn-cluster` containing all of the configuration and settings of your current cluster.

Reviewer: I'd add a point here showing that there's now a bunch of stuff being deployed by ArgoCD

Author: Done

The output of this command should be a string similar to "c-floral-wood-9134."

Let's now execute Commodore and compile the cluster.

[source,bash]
----
commodore catalog compile $CLUSTER_ID
----

Let's this time compile and push the cluster to GitLab.

[source,bash]
----
commodore catalog compile $CLUSTER_ID --push
----

At this point, there should be a GitLab project under your profile called `project-syn-cluster` containing all of the configuration and settings of your current cluster. And Argo CD automatically starts to synchronize settings and applications from the GitLab repository to the managed "k3d-steward" cluster.

=== Argo CD Status

If you have kept your Argo CD browser window open at http://localhost:35778 as explained in the "Accessing Argo CD" section of the xref:tutorials/getting-started.adoc[first part] of this tutorial, you would now see quite a bit of activity, with many applications deployed and featured in green color.

> 

Reviewer: Might be cool to note that once everything is synced (which should work now), you can access cluster metrics with `kubectl top nodes` and `kubectl top pods -A`.

Author: That's precisely the kind of cool things we must show at this stage. Will add it tomorrow, thanks!

Author: Done

= Getting Started with Project Syn: Working with Commodore

This document is the second part of the xref:tutorials/getting-started.adoc["Getting Started with Project Syn"] tutorial.

== Requirements

If you haven't yet, please execute the steps in the xref:tutorials/getting-started.adoc["Getting Started with Project Syn"] tutorial before attempting the instructions in this document.

This tutorial assumes that you have the "k3d-lieutenant" and "k3d-steward" clusters up and running and that you have installed all required tools.

In addition, create the following environment variable with the path of your private key:

[source,bash]
--
export COMMODORE_SSH_PRIVATE_KEY=~/.ssh/id_ed25519
--

Commodore requires this variable to be able to push changes to the GitLab repositories on your behalf.

== Commodore

> https://syn.tools/commodore/[Commodore] is the third component of the Project Syn suite of tools. It's a command-line tool that compiles the state of your clusters and pushes the changes to the GitLab projects. Steward watches those projects continuously and immediately applies the changes to your clusters, following the GitOps philosophy.

Reviewer 2: The changes are actually picked up by ArgoCD. Is this an intentional fuzziness?

Author: Again, it's a conceptual thing. What's the purpose of Steward if Argo CD does all the job? There must be a reason why Steward in there. Yes, in the background it is Argo doing the heavy work, but the crown of flowers goes to Steward. It's about giving conceptual responsibilities to these components at the level of a Getting Started guide.

Reviewer 2: Fair, I thought such thinking might be the case here.

Compile the information of your `k3d-steward` cluster using Commodore by running the command below, which will create a few environment variables in your shell.

[source,bash]
----
source <(curl -fsL https://try.syn.tools/commodore)
----

After the command above, you should be able to see a few variables and functions in your shell, for example, the Project Syn ID of your "k3d-steward" cluster:

[source,bash]
----
echo $CLUSTER_ID
----

Another thing you should see in your shell is a `commodore` command. Commodore is written in Python and requires quite a few external dependencies. To simplify this guide, we run Commodore as a container inside the "k3d-lieutenant" cluster, but you can install Commodore on its own in your system. Please refer to the xref:commodore::explanation/running-commodore.adoc["Running Commodore"] page in this documentation for details.

The output of this command should be a string similar to "c-floral-wood-9134."

Let's now execute Commodore and compile the cluster.

[source,bash]
----
commodore catalog compile $CLUSTER_ID
----

Let's this time compile and push the cluster to GitLab.

[source,bash]
----

> commodore catalog compile $CLUSTER_ID --push

Reviewer 2: Maybe compile only once using the `--interactive` option. This will show the output and push the changes in one go.

Author: All of these commands show lots of output. What other output is there to see?

Reviewer 2: Actually non, the difference is, with `--interactive --push` you only have to run compile once. One times less output.

----

> At this point, there should be a GitLab project under your profile called `project-syn-cluster` containing all of the configuration and settings of your current cluster. And Argo CD automatically starts to synchronize settings and applications from the GitLab repository to the managed "k3d-steward" cluster.

Reviewer 2: ```suggestion
At this point, there should be a GitLab project under your profile called `project-syn-cluster` containing all of the Kubernetes resouces for your current cluster. And Argo CD automatically starts to synchronize them from the GitLab repository to the managed "k3d-steward" cluster.
```

=== Argo CD Status

If you have kept your Argo CD browser window open at http://localhost:35778 as explained in the "Accessing Argo CD" section of the xref:tutorials/getting-started.adoc[first part] of this tutorial, you would now see quite a bit of activity, with many applications deployed and featuring "Healthy" and "Synced" statuses.

You can also see the activity in your cluster using the `kubectl top nodes` command:

[source,bash]
----
kubectl --context k3d-steward top nodes
----

The same with the `kubectl top pods -A` command:

[source,bash]
----
kubectl --context k3d-steward top pods -A
----

> 

Reviewer 2: Can we link to some follow-up documentations? Maybe pages explaining things in more detail?

The classic: Now what?

Author: It's in the works :) I'll add a link to those as soon as they are published.
